### PR TITLE
Final cleanup sprint

### LINKED ISSUE #33: Maybe remove direct dependency on jaxlib?
Jax does not depend on jaxlib.
It is not entirely clear to me why. 

I guess that it's because you want to be able to switch jaxlib version with/without CUDA support,
but our direct dependency should not hinder that, or does it?

### LINKED ISSUE #60: Respecting GPU Stream
Customcalls on gpu are passed a stream object in order to perform operations asynchronously.
Essentially they are supposed to enqueue operations to an operation buffer.

So we could assume that previous operations are also enqueued to the stream and not necessarily completed.

But our GPU operations do not respect this, and execute Syncronously (immediately), without checking that the stream of previous operations is empty.
I don't know if that is safe or not...

I believe that we should investigate whever we should change the code of `MPI_Allreduce` and similar to be more like
```python

cdef void mpi_allreduce(cudaStream_t* stream, void** buffers,
                        const char* opaque, size_t opaque_len) nogil except *:

    if COPY_TO_HOST:
        checked_cuda_memcpyAsync(in_buf, data, count, cudaMemcpyDeviceToHost, stream)

    cudaStreamSynchronize(stream)
    mpi_xla_bridge.mpi_allreduce(in_buf, out_buf, nitems, dtype, op, comm, token)

    if COPY_TO_HOST:
        # copy back to device
        checked_cuda_memcpy(out_data, out_buf, count, cudaMemcpyHostToDevice)
```

where the latter operation could be async, but then we need a callback (CudaEvent) to deallocate the cpu buffer when the memcpy is completed.

Also see https://on-demand.gputechconf.com/gtc/2014/presentations/S4236-multi-gpu-programming-mpi.pdf

However I'm not sure this is really needed. I never noticed data corruption so...

### PR DESCRIPTION
**Pull request by dionhaefner**
- [x] Use correct C types (`int` and `uintptr_t` instead of `int32_t` and `uint64_t`)
- [x] Refactor `utils.py`
- [x] Flush GPU stream before copy (fixes #60)
- [x] Ensure `setuptools_scm` is present
- [x] Runtime check for jaxlib version (fixes #33)
- [x] Clean up transpose logic in allreduce
- [x] Flush GPU only if it was used
- [x] Create separate default communicator
- [ ] <s>Raise Python exceptions on MPI error</s>
- [x] Do not raise Python exceptions from XLA bridge 
- [x] Raise error if omnistaging is disabled
- [x] Bump test coverage as much as possible
- [x] Make sure tests pass on all platforms

### CODE REVIEW

**Comment by PhilipVinc** [comment]
is this ready?

**Comment by dionhaefner** [comment]
Ready if you are.

File: docs/installation.rst [hunk 1]
```diff
@@ -4,29 +4,34 @@ Installation
...
+
+   $ pip install jax jaxlib
 
-   Much of the functionality we need has recently been added to JAX, which itself is changing frequently. Therefore, ``mpi4jax`` has somewhat strict requirements on the supported versions of JAX and jaxlib.
+.. note::
+
+   Much of the functionality we need has recently been added to JAX, which itself is changing frequently. Therefore, ``mpi4jax`` has somewhat strict requirements on the supported versions of JAX and jaxlib. Be prepared to upgrade!
 
-We recommend that you use ``pip`` to install mpi4jax, however a distribution is also available on ``conda``:
+We recommend that you use ``pip`` to install mpi4jax (but a distribution is also available via ``conda``):
```
Comment (PhilipVinc) [inline docs/installation.rst:17]:
```suggestion
We recommend that you use ``pip`` to install mpi4jax (but a distribution is also available via ``conda`` which will work if MPI, mpi4py and mpi4jax are all installed through conda ):
```

Comment (dionhaefner) [reply docs/installation.rst:17]: How is this? I think it is a common best practice that you shouldn't mix conda and pip. Conda users hopefully already know this.

File: mpi4jax/_src/xla_bridge/__init__.py [hunk 2]
```diff
@@ -13,46 +9,10 @@
...
-        has_cuda_mpi = True
-    elif is_falsy(gpu_copy_behavior):
-        has_cuda_mpi = False
-    else:
-        has_cuda_mpi = False
-        warn_msg = (
-            "Not using CUDA-enabled MPI. "
-            "If you are sure that your MPI library is built with CUDA support, "
-            "set MPI4JAX_USE_CUDA_MPI=1. To silence this warning, "
-            "set MPI4JAX_USE_CUDA_MPI=0."
```
Comment (PhilipVinc) [inline mpi4jax/_src/xla_bridge/__init__.py:42]: You moved the warning somewhere else or you removed it?

Comment (dionhaefner) [reply mpi4jax/_src/xla_bridge/__init__.py:42]: It's here now:

https://github.com/PhilipVinc/mpi4jax/pull/62/files#diff-20bdfe84be6f1fd1effb18bb019f53d6a47fee9adb3a039a062899f775bcc191R65-R89

**Review by PhilipVinc — APPROVED** [review]

**Pull request merged**